amethyst_ui/src/pass.rs

```diff
@@ -0,0 +1,274 @@
+            if cached_draw_order
+                .iter()
+                .position(|&(_z, cached_entity)| entity == cached_entity)
+                .is_none()
+            {
+                let pos = cached_draw_order
+                    .iter()
+                    .position(|&(cached_z, _)| transform.z >= cached_z);
+                match pos {
+                    Some(pos) => cached_draw_order.insert(pos, (transform.z, entity)),
+                    None => cached_draw_order.push((transform.z, entity)),
+                }
+            }
+        }
+
+        // Sort from largest z value to smallest z value.
+        cached_draw_order.sort_unstable_by(|&(z1, _), &(z2, _)| {
+            z2.partial_cmp(&z1).unwrap_or(Ordering::Equal)
+        });
+
+        // This pass can't be executed in parallel, so we use a dumby bitset of a
+        // single element to provide a fake parallel iterator that performs the entire
+        // pass in the first iteration.
+        let mut bitset = BitSet::new();
+        bitset.add(0);
+
+        let cached_draw_order = &cached_draw_order;
+
+        let proj_vec = vec2(
+            2. / screen_dimensions.width(),
+            -2. / screen_dimensions.height(),
+        ).extend(-2.)
+            .extend(1.);
+
+        supplier
+            .supply(bitset.par_join().map(move |_id| {
+                move |encoder: &mut Encoder, effect: &mut Effect| for &(_z, entity) in
+                    cached_draw_order.iter()
+                {
+                    // These are safe as we guaranteed earlier these entities are present.
+                    let ui_transform = ui_transform.get(entity).unwrap();
+                    if let Some(mesh) = mesh_storage.get(unit_mesh) {
+                        let vbuf = match mesh.buffer(PosTex::ATTRIBUTES) {
+                            Some(vbuf) => vbuf.clone(),
+                            None => continue,
+                        };
+                        let vertex_args = VertexArgs {
+                            proj_vec: proj_vec.into(),
+                            coord: [ui_transform.x, ui_transform.y],
+                            dimension: [ui_transform.width, ui_transform.height],
+                        };
+                        effect.update_constant_buffer("VertexArgs", &vertex_args, encoder);
+                        effect.data.vertex_bufs.push(vbuf);
+                        if let Some(image) = ui_image
+                            .get(entity)
+                            .and_then(|image| tex_storage.get(&image.texture))
+                        {
+                            effect.data.textures.push(image.view().clone());
+                            effect.data.samplers.push(image.sampler().clone());
+                            effect.draw(mesh.slice(), encoder);
+                            effect.clear();
+                        }
+
+                        if let Some(image) = ui_text
```

Comment:
nit: `image` should be `text`

Reply:
I feel it's important in this context to indicate we're drawing an image rendered by another system, that's why I chose image.